profiler/profile.go

```diff
@@ -149,18 +148,49 @@ var profileTypes = map[ProfileType]profileType{
 	MetricsProfile: {
 		Name:     "metrics",
 		Filename: "metrics.json",
-		Collect: func(_ profileType, p *profiler) ([]byte, error) {
+		Collect: func(p *profiler) ([]byte, error) {
 			var buf bytes.Buffer
 			err := p.met.report(now(), &buf)
 			return buf.Bytes(), err
 		},
 	},
 }
 
-func collectGenericProfile(t profileType, p *profiler) ([]byte, error) {
-	var buf bytes.Buffer
-	err := p.lookupProfile(t.Name, &buf, 0)
-	return buf.Bytes(), err
+func collectGenericProfile(name string, delta *pprofutils.Delta) func(p *profiler) ([]byte, error) {
+	return func(p *profiler) ([]byte, error) {
+		var extra []*pprofile.Profile
+		if cAlloc, ok := extensions.GetCAllocationProfiler(); ok && p.cfg.deltaProfiles && name == "heap" {
+			// For the heap profile, we'd also like to include C
+			// allocations if that extension is enabled and have the
+			// allocations show up in the same profile. Collect them
+			// first before getting the regular heap snapshot so
+			// that all allocations cover the same time period
+			//
+			// TODO: Support non-delta profiles for C allocations?
+			cAlloc.Start(2 * 1024 * 1024)
+			p.interruptibleSleep(p.cfg.period)
+			profile, err := cAlloc.Stop()
+			if err == nil {
+				extra = append(extra, profile)
+			}
+		}
```

Comment:
Maybe the CAlloc logic could be implemented in the HeapProfile profileType via an additional func. E.g.

```go
if pt.CollectExtra != nil {
  extra = append(extra, pt.CollectExtra())
}
```

Something like that. Of course it will require bringing the `pt profileType` argument back.